lib/active_merchant/billing/gateways/credorax.rb

```diff
@@ -0,0 +1,202 @@
+        reference_action = add_reference(post, authorization)
+        add_customer_data(post, options)
+
+        commit(:refund, post)
+      end
+
+      def credit(amount, payment_method, options={})
+        post = {}
+        add_invoice(post, amount, options)
+        add_payment_method(post, payment_method)
+        add_customer_data(post, options)
+
+        commit(:credit, post)
+      end
+
+      def verify(credit_card, options={})
+        MultiResponse.run(:use_first_response) do |r|
+          r.process { authorize(100, credit_card, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((b1=)\d+), '\1[FILTERED]').
+          gsub(%r((b5=)\d+), '\1[FILTERED]')
+      end
+
+      private
+
+      def add_invoice(post, money, options)
+        post[:a4] = amount(money)
+        post[:a1] = options[:order_id] || generate_unique_id
+        post[:a5] = options[:currency] || currency(money)
+      end
+
+      CARD_TYPES = {
+        "visa" => '1',
+        "mastercard" => '2',
+        "maestro" => '9'
+      }
+
+      def add_payment_method(post, payment_method)
+        post[:c1] = payment_method.name
+        post[:b2] = CARD_TYPES[payment_method.brand] || '0'
+        post[:b1] = payment_method.number
+        post[:b5] = payment_method.verification_value
+        post[:b4] = format(payment_method.year, :two_digits)
+        post[:b3] = format(payment_method.month, :two_digits)
+      end
+
+      def add_customer_data(post, options)
+        post[:d1] = options[:ip] || '127.0.0.1'
+        post[:c3] = options[:email] || 'unspecified@example.com'
+        if (billing_address = options[:billing_address])
+          post[:c5] = billing_address[:address1]+ billing_address[:address2]
+          post[:c7] = billing_address[:city]
+          post[:c10]    = billing_address[:zip]
+          post[:c2] = billing_address[:phone]
+        end
+      end
+
+      def add_reference(post, authorization)
+        response_id, authorization_code, request_id, action = authorization.split(";")
+        post[:g2] = response_id
+        post[:g3] = authorization_code
+        post[:g4] = request_id
+        action || :authorize
+      end
+
+      ACTIONS = {
+        purchase: '1',
+        authorize: '2',
+        capture: '3',
+        authorize_void:'4',
+        refund: '5',
+        credit: '6',
+        purchase_void: '7',
+        capture_void: '9'
+      }
+
+      def commit(action, params, reference_action = nil)
+        raw_response = ssl_post(url, post_data(action, params, reference_action))
+        response = parse(raw_response)
+
+        Response.new(
+          success_from(response),
+          message_from(response),
+          response,
+          authorization: "#{response["Z1"]};#{response["Z4"]};#{response["A1"]};#{action}",
```

Comment:
ugh - all of these are required to properly correlate future reference txs? :crying_cat_face: